targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c

```diff
@@ -0,0 +1,100 @@
+/* mbed Microcontroller Library
+ * (C)Copyright TOSHIBA ELECTRONIC DEVICES & STORAGE CORPORATION 2017 All rights reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "mbed_error.h"
+#include "tmpm46B_esg.h"
+#include "trng_api.h"
+
+volatile uint32_t CompleteFlag = 0U; // Global variables
+
+static void ESG_Config(void)
+{
+    uint16_t latchtiming;
+    uint16_t Fintming = 560U;
+    // Confirm the ESG core stops
+    if (ESG_GetCalculationStatus() == ESG_CALCULATION_COMPLETE) {
+        // Confirm no interrupt generation
+        if (ESG_GetIntStatus() == DISABLE) {
+            // Set the latch timing
+            ESG_SetLatchTiming(ESG_LATCH_TIMING_1);
+            // Get latchtiming value
+            latchtiming = ESG_GetLatchTiming();
+            if (Fintming >= 512U * (latchtiming + 1U) + 3U) {
+                // Set the output timing
+                ESG_SetFintiming(Fintming);
+            } else {
+                error("Fintming value input error !\n");
+            }
+        } else {
+            // Do nothing
+        }
+    } else {
+        // Do nothing 
+    }
+}
+
+void trng_init(trng_t *obj)
+{
+    (void)obj;
+
+    // Enable clock supply
+    CG_SetFcPeriphB(CG_FC_PERIPH_ESG,ENABLE);
+    // Enable interrupt
+    NVIC_ClearPendingIRQ(INTESG_IRQn);
+    NVIC_EnableIRQ(INTESG_IRQn);
+    // Initialise ESG core
+    ESG_Config();
+    // Start up the ESG core
+    ESG_Startup();
+}
+
+void trng_free(trng_t *obj)
+{
+    (void)obj;
+    // Disable interrupt
+    NVIC_DisableIRQ(INTESG_IRQn);
+    // Disable clock supply
+    CG_SetFcPeriphB(CG_FC_PERIPH_ESG,DISABLE);
+}
+
+int trng_get_bytes(trng_t *obj, uint8_t *output, size_t length, size_t *output_length)
+{
+    uint32_t Data[length];
+    int timeout = 10;
```

Comment:
how is  t his number calculated ? why 10?

Reply:
The entropy seed generator (ESG) is a circuit that generates a 512-bit entropy seed using a ring oscillator. When the ESG core ends the operation, it outputs an interrupt. Interrupt generation can be checked by polling of the interrupt flag. Then an entropy seed can be read from 16 blocks (ESGBLK00 to 15)
Timeout 10 is used for 10 iteration to confirm the ESG operation is completed.
- Sometimes function execution completes before generating interrupts and read wrong value.
- Timeout will help to return from function, even interrupts are not generated within 10 iteration.

We will also try to change the approach.

Reply:
@hanno-arm Please review trng addition